File: 0001~0500/0342. Power of Four/main.py
"""
Source: https://rb.gy/3wep8
Date: 2023/8/21
Skill:
Ref:
Runtime: 42 ms, faster than 69.21%
Memory Usage: 16.32 MB, less than 26.51%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right
from functools import cache


class Solution:
    def isPowerOfFour(self, n: int) -> bool:
        if n <= 0:
            return False
        num = 1
        while num < n:
            num <<= 2
        return num == n


if __name__ == "__main__":
    s = Solution()

File: 0001~0500/0342. Power of Four/main.cpp
/**
 * Source: tinyurl.com/2xa87ucn
 * Date: 2023/10/23
 * Skill:
 * Runtime: 0 ms, faster than 100.00% of C++ online submissions
 * Memory Usage: 6.30 MB, less than 47.70% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
public:
    bool isPowerOfFour(int n) {
        if (n == 1)
            return true;
        if (n < 1 || n % 4 != 0)
            return false;
        return isPowerOfFour(n >> 2);
    }
};


int main() {
    Solution s;
    vector<int> nums{89135};
    vector<vector<int>> arr{{1,5}, {10,11},{12,18},{20,25},{30,32}};
    function<int(int, int)> test;
    test = [&](int a, int b) {
        if (a > 0)
            return test(a - 1, 0);
        else
            return 0;
    };
}